src/cursor/cur_version.c

```diff
@@ -0,0 +1,546 @@
+                F_SET(version_cursor, WT_VERSION_CUR_ON_DISK_EXHAUSTED);
+                F_SET(version_cursor, WT_VERSION_CUR_HS_EXAUSTED);
+                WT_ERR(WT_NOTFOUND);
+            }
+            break;
+        case WT_PAGE_COL_FIX:
+            /*
+             * If search returned an insert, we might be past the end of page in the append list, so
+             * there's no on-disk value.
+             */
+            if (cbt->recno >= cbt->ref->ref_recno + page->entries) {
+                F_SET(version_cursor, WT_VERSION_CUR_ON_DISK_EXHAUSTED);
+                F_SET(version_cursor, WT_VERSION_CUR_HS_EXAUSTED);
+                WT_ERR(WT_NOTFOUND);
+            }
+            break;
+        case WT_PAGE_COL_VAR:
+            /* Empty page doesn't have any on page value. */
+            if (page->entries == 0) {
+                F_SET(version_cursor, WT_VERSION_CUR_ON_DISK_EXHAUSTED);
+                F_SET(version_cursor, WT_VERSION_CUR_HS_EXAUSTED);
+                WT_ERR(WT_NOTFOUND);
+            }
+            break;
+        default:
+            WT_ERR(__wt_illegal_value(session, page->type));
+        }
+
+        /* Get the ondisk value. */
+        WT_ERR(__wt_value_return_buf(cbt, cbt->ref, &cbt->upd_value->buf, &cbt->upd_value->tw));
+
+        if (!WT_TIME_WINDOW_HAS_STOP(&cbt->upd_value->tw)) {
+            durable_stop_ts = version_cursor->upd_durable_stop_ts;
+            stop_ts = version_cursor->upd_stop_ts;
+            stop_txn = version_cursor->upd_stop_txnid;
+        } else {
+            durable_stop_ts = cbt->upd_value->tw.durable_stop_ts;
+            stop_ts = cbt->upd_value->tw.stop_ts;
+            stop_txn = cbt->upd_value->tw.stop_txn;
+        }
+
+        __wt_cursor_set_value(cursor, cbt->upd_value->tw.start_txn, cbt->upd_value->tw.start_ts,
+          cbt->upd_value->tw.durable_start_ts, stop_txn, stop_ts, durable_stop_ts,
+          WT_UPDATE_STANDARD, cbt->upd_value->tw.prepare, 0, WT_VERSION_DISK_IMAGE);
+
+        upd_found = true;
+        F_SET(version_cursor, WT_VERSION_CUR_ON_DISK_EXHAUSTED);
+    }
+
+    if (!upd_found && !F_ISSET(version_cursor, WT_VERSION_CUR_HS_EXAUSTED)) {
+        /* Ensure we can see all the content in the history store. */
+        F_SET(hs_cursor, WT_CURSTD_HS_READ_COMMITTED);
+
+        if (!F_ISSET(hs_cursor, WT_CURSTD_KEY_INT)) {
+            /* Ensure enough room for a column-store key without checking. */
+            WT_ERR(__wt_scr_alloc(session, WT_INTPACK64_MAXSIZE, &key));
+
+            if (page->type == WT_PAGE_ROW_LEAF) {
+                key->data = table_cursor->key.data;
+                key->size = table_cursor->key.size;
+            } else {
+                p = key->mem;
+                WT_ERR(__wt_vpack_uint(&p, 0, cbt->recno));
+                key->size = WT_PTRDIFF(p, key->data);
+            }
+            hs_cursor->set_key(hs_cursor, 4, S2BT(session)->id, key, WT_TS_MAX, UINT64_MAX);
+            __wt_scr_free(session, &key);
```

Comment:
@keithbostic If I move this line of the code to the err section, it will crash there if compiled with gcc 8.2, which is the version used in evergreen. Moving this line here passes the linux tests but the windows tests are still crashing. On my local box I build with gcc 7.5 and everything is fine. I'm not sure what I'm doing wrong here with the key scratch buffer. Basically, I refer to the code in hs_rec which does the same thing:

```
        /* History store table key component: source key. */
        switch (r->page->type) {
        case WT_PAGE_COL_FIX:
        case WT_PAGE_COL_VAR:
            p = key->mem;
            WT_ERR(__wt_vpack_uint(&p, 0, WT_INSERT_RECNO(list->ins)));
            key->size = WT_PTRDIFF(p, key->data);
            break;
        case WT_PAGE_ROW_LEAF:
            if (list->ins == NULL) {
                WT_WITH_BTREE(
                  session, btree, ret = __wt_row_leaf_key(session, r->page, list->rip, key, false));
                WT_ERR(ret);
            } else {
                key->data = WT_INSERT_KEY(list->ins);
                key->size = WT_INSERT_KEY_SIZE(list->ins);
            }
            break;
        default:
            WT_ERR(__wt_illegal_value(session, r->page->type));
        }
```

Reply:
I don't see anything obvious, I'm afraid -- have you run with a sanitizer build? (My bet is that will tag something.) I'm happy to look further if you can tell me what test is failing?

Reply:
@quchenhao, see my comment above, it looks like the crashes are all on row stores.

Reply:
I'm confident there's something obvious I'm missing, but in the case of RS, we're either assigning `data/size` or calling a function that will allocate more buffer space as necessary?  Shouldn't both of those be safe?

Reply:
You're right Keith.  Perhaps the problem is:
```
               key->data = table_cursor->key.data;
                key->size = table_cursor->key.size;
```
We're re-assigning the scratch buffer mem to be something else, and then there are problem when it is freed.  The row store path shouldn't use the scratch buffer.

Reply:
I ran with a sanitizer build; cursor17 runs, but cursor18 and cursor19 both fail with this:
```
% sanpyt test_cursor18
.........=================================================================
==1817120==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffcd29abf30 at pc 0x7f21604fa0aa bp 0x7ffcd29aa670 sp 0x7ffcd29aa668
WRITE of size 8 at 0x7ffcd29abf30 thread T0
    #0 0x7f21604fa0a9 in __wt_struct_unpackv /home/ubuntu/wt.c/./src/include/packing_inline.h:718:9
    #1 0x7f2160504a25 in __wt_cursor_get_valuev /home/ubuntu/wt.c/src/cursor/cur_std.c:542:15
    #2 0x7f216050005a in __wt_cursor_get_value /home/ubuntu/wt.c/src/cursor/cur_std.c:507:11
    #3 0x7f21605ba894 in __curversion_next_int /home/ubuntu/wt.c/src/cursor/cur_version.c:278:9
    #4 0x7f21605a5382 in __curversion_next /home/ubuntu/wt.c/src/cursor/cur_version.c:329:5
    #5 0x7f2161b36c7a in _wrap_Cursor_next /home/ubuntu/wt.c/lang/python/wiredtiger_wrap.c:3837:16
    #6 0x7f21657408f1 in cfunction_vectorcall_O /data/mci/7e82c9d00383110b1bd32eed42456241/toolchain-builder/tmp/build-python.sh-NPI/build-Python-3.9.2/../src/Python-3.9.2/Objects/methodobject.c:512
    #7 0x7f2165771e27 in _PyObject_VectorcallTstate /data/mci/7e82c9d00383110b1bd32eed42456241/toolchain-builder/tmp/build-python.sh-NPI/build-Python-3.9.2/../src/Python-3.9.2/Include/cpython/abstract.h:118:11

Address 0x7ffcd29abf30 is located in stack of thread T0 at offset 176 in frame
    #0 0x7f21605b51cf in __curversion_next_int /home/ubuntu/wt.c/src/cursor/cur_version.c:94

  This frame has 5 object(s):
    [32, 40) 'key' (line 98)
    [64, 104) 'hs_value' (line 100)
    [144, 152) 'twp' (line 102)
    [176, 180) 'hs_upd_type' (line 106) <== Memory access at offset 176 partially overflows this variable
    [192, 200) 'p' (line 108)
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow /home/ubuntu/wt.c/./src/include/packing_inline.h:718:9 in __wt_struct_unpackv
```
I think `uint32_t hs_upd_type` needs to be a `uint64_t`.

This is close to the lines you asked about, so this might be the problem.

Reply:
Thanks @keithbostic @ddanderson

Reply:
@keithbostic @ddanderson The test is still crashing with the integer overflow fix. There is no useful information in the windows crash. From my understanding, the set key api will copy the key and pack it into its own key buffer. So it should be safe to pass it to the set_key api.